### PR TITLE
Adding updateDob funtionality for TIs

### LINKED ISSUE #3276: Create "Update Dob" funtionality for the TIs to udate their clients Date of birth
Currently TIs don't have the ability to update their Client's Date of Birth as the old way of code did not persist the Date of birth field. This feature will have the ability to directly edit the DOB and also persists the new data.

### PR DESCRIPTION
**Pull request by swatkat1**
### Description

This feature provides a way for TIs to add/update the Date of birth of their Clients directly in the TI dashboard. 

## Release notes:
The Update DOB functionality provides a way for Trusted Intermediaries to Add/Edit the Dob column values of TI clients directly in the TI Dashboard.

### Checklist

- [x] Added the correct label: < feature | bug | dependencies | infrastructure | ignore-for-release >
- [x] Created tests which fail without the change (if possible)
- [x] Extended the README / documentation, if necessary

### Issue(s)

Fixes #3276 

### Screenshots

Display: Note- the Test Applicant's DOB is 08/01/2021.

image

After Update: Note- the Test Applicant's DOB is changed to 07/21/1941.

image

### CODE REVIEW

File: server/app/repository/UserRepository.java [hunk 1]
```diff
@@ -250,7 +250,7 @@ public void removeTrustedIntermediaryFromGroup(long id, long accountId)
     }
   }
 
-  private Optional<Account> lookupAccount(long accountId) {
+  public Optional<Account> lookupAccount(long accountId) {
```
Comment (clouser2) [inline server/app/repository/UserRepository.java:253]: See the comment in TrustedIntermediaryService. If this isn't used, I think we can change the visibility back to private.

File: server/test/controllers/ti/TrustedIntermediaryControllerTest.java [hunk 2]
```diff
@@ -50,7 +52,7 @@ public void addApplicant_WithMissingDob() {
                         "dob",
                         "")));
     TrustedIntermediaryGroup group = repo.listTrustedIntermediaryGroups().get(0);
-    Result result = tiController.addApplicant(group.id + 1, requestBuilder.build());
+    Result result = tiController.addApplicant(group.id + 2, requestBuilder.build());
```
Comment (clouser2) [inline server/test/controllers/ti/TrustedIntermediaryControllerTest.java:55]: Can you help me understand why this change was necessary? Is there anyway we can avoid the need for a magic number here?

File: server/test/services/ti/TrustedIntermediaryServiceTest.java [hunk 3]
```diff
@@ -300,4 +303,40 @@ private void setupTIAccount(
...
+    Form<UpdateApplicantDob> form =
+        formFactory.form(UpdateApplicantDob.class).bindFromRequest(requestBuilder.build());
+    Account account = repo.lookupAccountByEmail("email30").get();
+    Form<UpdateApplicantDob> returnedForm = service.updateApplicantDateOfBirth(account.id, form);
+    assertThat(returnedForm.hasErrors()).isTrue();
+    assertThat(returnedForm.error("dob").get().message())
+        .isEqualTo("Date of Birth must be in MM/dd/yyyy format");
+  }
+
+  @Test
```
Comment (clouser2) [inline server/test/services/ti/TrustedIntermediaryServiceTest.java:330]: Mind adding a test for the case called out above where the DOB is attempted to be updated for an account not tied to the trusted intermediary?

**Review by clouser2 — CHANGES_REQUESTED** [review]

File: browser-test/src/support/ti_dashboard.ts [hunk 4]
```diff
@@ -26,6 +27,25 @@ export class TIDashboard {
...
+      )}")`,
+    )
+    expect(`.cf-admin-question-table-row:has-text("${client.emailAddress}")`)
+    expect(`.cf-admin-question-table-row:has-text("${client.firstName}")`)
+    expect(`.cf-admin-question-table-row:has-text("${client.lastName}")`)
+  }
+
+  async updateClientDateOfBirth(client: ClientInformation, newDobDate: string) {
+    await this.page.locator('id=date-of-birth-update').fill(newDobDate)
+    await this.page.click('text ="Update DOB"')
```
[the remark this answers is not in the document]
Comment (swatkat1) [reply browser-test/src/support/ti_dashboard.ts:46]: done!

File: server/app/controllers/ti/TrustedIntermediaryController.java [hunk 5]
```diff
@@ -11,19 +11,23 @@
...
 import play.data.Form;
 import play.data.FormFactory;
+import play.data.validation.ValidationError;
 import play.i18n.MessagesApi;
 import play.mvc.Http;
 import play.mvc.Result;
 import repository.SearchParameters;
 import repository.UserRepository;
 import services.PaginationInfo;
+import services.applicant.exception.ApplicantNotFoundException;
```
[the remark this answers is not in the document]
Comment (swatkat1) [reply server/app/controllers/ti/TrustedIntermediaryController.java:30]: Done!

File: server/app/controllers/ti/TrustedIntermediaryController.java [hunk 6]
```diff
@@ -100,6 +104,40 @@ public Result dashboard(
...
+
+    Optional<TrustedIntermediaryGroup> trustedIntermediaryGroup =
+        userRepository.getTrustedIntermediaryGroup(civiformProfile.get());
+    if (trustedIntermediaryGroup.isEmpty()) {
+      return notFound();
+    }
+
+    Form<UpdateApplicantDob> form =
+        formFactory.form(UpdateApplicantDob.class).bindFromRequest(request);
+    Form<UpdateApplicantDob> returnedForm = null;
```
[the remark this answers is not in the document]
Comment (swatkat1) [reply server/app/controllers/ti/TrustedIntermediaryController.java:122]: Done! thank you!

File: server/app/controllers/ti/TrustedIntermediaryController.java [hunk 7]
```diff
@@ -11,19 +11,23 @@
 import com.google.common.base.Preconditions;
 import controllers.BadRequestException;
 import forms.AddApplicantToTrustedIntermediaryGroupForm;
+import forms.UpdateApplicantDob;
```
[the remark this answers is not in the document]
Comment (swatkat1) [reply server/app/controllers/ti/TrustedIntermediaryController.java:14]: Done!

File: server/app/controllers/ti/TrustedIntermediaryController.java [hunk 8]
```diff
@@ -100,6 +104,40 @@ public Result dashboard(
...
+    }
+
+    Form<UpdateApplicantDob> form =
+        formFactory.form(UpdateApplicantDob.class).bindFromRequest(request);
+    Form<UpdateApplicantDob> returnedForm = null;
+
+    try {
+      returnedForm = tiService.updateApplicantDateOfBirth(accountId, form);
+    } catch (ApplicantNotFoundException e) {
+      throw new BadRequestException(e.getMessage());
```
[the remark this answers is not in the document]
Comment (swatkat1) [reply server/app/controllers/ti/TrustedIntermediaryController.java:127]: Actually this exception is used quite a bit in the Applicant Service, Applicant Repository and Applicant**Controllers... I am not very sure how the errors should be handled there if this exception is thrown. That would be a little beyond the scope of this PR. Other than modifying how this error is handled across the code base, i would throw 404 than 400.

File: server/app/controllers/ti/TrustedIntermediaryController.java [hunk 9]
```diff
@@ -100,6 +104,40 @@ public Result dashboard(
...
+    }
+
+    if (!returnedForm.hasErrors()) {
+      return redirect(
```
[the remark this answers is not in the document]
Comment (swatkat1) [reply server/app/controllers/ti/TrustedIntermediaryController.java:131]: Done1 thank you .. learnt something new here :+1:

File: server/app/controllers/ti/TrustedIntermediaryController.java [hunk 10]
```diff
@@ -100,6 +104,40 @@ public Result dashboard(
...
+          routes.TrustedIntermediaryController.dashboard(
```
[the remark this answers is not in the document]
Comment (swatkat1) [reply server/app/controllers/ti/TrustedIntermediaryController.java:132]: I think so...But good call.. I will follow up with Elise..

File: server/app/forms/UpdateApplicantDob.java [hunk 11]
```diff
@@ -0,0 +1,13 @@
+package forms;
+
+public final class UpdateApplicantDob {
```
[the remark this answers is not in the document]
Comment (swatkat1) [reply server/app/forms/UpdateApplicantDob.java:3]: Sure!

File: server/app/services/ti/TrustedIntermediaryService.java [hunk 12]
```diff
@@ -159,4 +172,43 @@ private ImmutableList<Account> searchAccounts(
...
+   */
+  public Form<UpdateApplicantDob> updateApplicantDateOfBirth(
+      Long accountId, Form<UpdateApplicantDob> form) throws ApplicantNotFoundException {
+
+    form = validateDateOfBirthForUpdateDob(form);
+
+    if (form.hasErrors()) {
+      return form;
+    }
+    Optional<Account> optionalAccount = userRepository.lookupAccount(accountId);
```
[the remark this answers is not in the document]
Comment (swatkat1) [reply server/app/services/ti/TrustedIntermediaryService.java:196]: There is already a method which does this, so why implement a another one? Also, the TIs cannot update the ther TIs applicant information as it won't be visible to them.  The dashboard() in the controllers, gets the account only managed by that TI and allows them to edit/view the information of their clients.

File: server/test/controllers/ti/TrustedIntermediaryControllerTest.java (hunk 2 above)
Comment (swatkat1) [reply server/test/controllers/ti/TrustedIntermediaryControllerTest.java:55]: I can't :(

File: server/test/services/ti/TrustedIntermediaryServiceTest.java (hunk 3 above)
Comment (swatkat1) [reply server/test/services/ti/TrustedIntermediaryServiceTest.java:330]: I removed that check as that case is kind of impossible. I would like to pick your brains, if you think its possible.

File: browser-test/src/support/ti_dashboard.ts [hunk 13]
```diff
@@ -26,6 +27,25 @@ export class TIDashboard {
     await this.page.click('text="Add"')
   }
 
+  async expectClientDateOfBirthUpdated(
+    client: ClientInformation,
+    newDob: string,
+  ) {
+    expect(
```
[the remark this answers is not in the document]
Comment (swatkat1) [reply browser-test/src/support/ti_dashboard.ts:34]: Ok.. nice optimization

File: server/app/repository/UserRepository.java (hunk 1 above)
Comment (swatkat1) [reply server/app/repository/UserRepository.java:253]: Done!

File: browser-test/src/support/ti_dashboard.ts [hunk 14]
```diff
@@ -2,7 +2,8 @@ import {Page} from 'playwright'
 
 /*
  * This class is to test Civiform in the TI path
- * It impements functionality to create a new client,Test if the given client is present in the dashboard
+ * It impements functionality to create a new client,Test if the given client is present/not present in the dashboard
+ * It can also update a Client's Data of Birth to a new one
```
Comment (syntapy) [inline browser-test/src/support/ti_dashboard.ts:6]: type?

File: server/test/services/ti/TrustedIntermediaryServiceTest.java (hunk 3 above)
Comment (swatkat1) [reply server/test/services/ti/TrustedIntermediaryServiceTest.java:348]: Redone as we discussed yesterday. Testcase: updateApplicantDateOfBirth_throwsApplicantNotFoundExceptionDueToIncorrectTIGroup is added.

File: server/app/controllers/ti/TrustedIntermediaryController.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (clouser2) [reply server/app/controllers/ti/TrustedIntermediaryController.java]: As I understand, all other callers would have to explicitly catch the exception since it's a checked exception. As such, adding it to ErrorHandler would mean that it would continue to work as intended here and we could slowly remove the explicit catches in other places that use it (similar to what we've been doing for ProgramNotFoundException).

I believe this should be safe to add to ErrorHandler without affecting the other callsites.

File: server/test/services/DateConverterTest.java [hunk 15]
```diff
@@ -38,7 +38,7 @@ public void parseIso8601DateToLocalDate_isSuccessful() {
 
   @Test
   public void renderDate_isSuccessful() {
-    String expectedResult = "2020/01/01";
+    String expectedResult = "2020-01-01";
```
Comment (clouser2) [inline server/test/services/DateConverterTest.java:41]: Will this change cause other date renderings to change in the UI? If so, are we sure that dashes are what we want rather than slashes in those cases?

I'd like to better understand why this is necessary, especially since all of the screenshots I see use "/".

**Review by clouser2 — CHANGES_REQUESTED** [review]
All looks good modulo one question around the changes to DateConverter.

File: server/test/services/DateConverterTest.java (hunk 15 above)
Comment (swatkat1) [reply server/test/services/DateConverterTest.java:41]: https://stackoverflow.com/questions/14212527/how-to-set-default-value-to-the-inputtype-date -- I knew you would ask this :).. also i checked the code-base, this is not used anywhere, hence could safely change it.

File: server/test/controllers/ti/TrustedIntermediaryControllerTest.java (hunk 2 above)
Comment (swatkat1) [reply server/test/controllers/ti/TrustedIntermediaryControllerTest.java]: this is fixed.. no magic number needed

File: server/app/controllers/ti/TrustedIntermediaryController.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (swatkat1) [reply server/app/controllers/ti/TrustedIntermediaryController.java]: Done! Thank you

File: browser-test/src/support/ti_dashboard.ts (hunk 14 above)
Comment (swatkat1) [reply browser-test/src/support/ti_dashboard.ts:6]: @syntapy - Can you please approve the PR if you don't have further comments?

File: server/app/services/ti/TrustedIntermediaryService.java [hunk 16]
```diff
@@ -159,4 +172,50 @@ private ImmutableList<Account> searchAccounts(
...
+    if (form.hasErrors()) {
+      return form;
+    }
+    Optional<Account> optionalAccount =
+        trustedIntermediaryGroup.getManagedAccounts().stream()
+            .filter(account -> account.id.equals(accountId))
+            .findAny();
+
+    if (optionalAccount.isEmpty() || optionalAccount.get().newestApplicant().isEmpty()) {
+      throw new RuntimeException(new ApplicantNotFoundException(accountId));
```
Comment (clouser2) [inline server/app/services/ti/TrustedIntermediaryService.java:205]: Why a runtime exception? Can this throw ApplicantNotFoundException directly?

When that happens, the method will need to be declared to throw ApplicantNotFoundException. Then the top-level method in the controller `updateDateOfBirth` will also need to also declare that ApplicantNotFoundException is thrown.

**Review by clouser2 — CHANGES_REQUESTED** [review]

File: server/test/services/DateConverterTest.java (hunk 15 above)
Comment (swatkat1) [reply server/test/services/DateConverterTest.java:41]: It is taking a LocalDate...

Comment (clouser2) [reply server/test/services/DateConverterTest.java:41]: Correct. I'm suggest naming the method `formatIso8601Date` with a param of `LocalDate`.

File: server/app/services/ti/TrustedIntermediaryService.java (hunk 16 above)
Comment (swatkat1) [reply server/app/services/ti/TrustedIntermediaryService.java]: Changed it back to ApplicantNotFoundException.

**Review by clouser2 — APPROVED** [review]
